### DISCUSSION TITLE
Migrations using postgres_fdw

### DISCUSSION THREAD

**Discussion by jkepps** [prisma/prisma#8610] [Q&A]
Is it possible to reference foreign schemas through the postgres_fdw extension in prisma migrations?

In my use case I created a server with the postgres_fdw extension, then imported a foreign schema from that server and can now access it as if it was in my original server. Heroku provides this functionality out of the box with [heroku data-links](https://devcenter.heroku.com/articles/heroku-data-links) so I'm hoping this isn't too uncommon of a use case.

When I try to access data from the imported foreign schema within a prism migration, I get an error with the message `The underrlying table for model 'datalink_platform.patients' does not exist.` Where `datalink_platform` is the name of the linked schema that points to a foreign datastore. I am able to use this syntax within my server code to access the `patients` table, so I'm fairly certain this is due to the internal workings of the prisma migration system.

**Comment by janpio** [comment]
We have not encountered that use case yet (so it might not be _that_ common), so it not working would not be surprising. But it sounds pretty interesting.

Some context:
https://www.postgresql.org/docs/9.5/postgres-fdw.html
https://www.postgresql.org/docs/9.5/sql-importforeignschema.html

What do you mean by "access data from the imported foreign schema within a prisma migration"?

Which service tiers on Heroku is this available out of the box? Can you help us quickly set up a reproduction by typing up some instructions on how to configure this? We would really like to try this.

**Reply by janpio** [reply]
(Created https://github.com/prisma/prisma/issues/8614 for our internal tracking)

**Reply by jkepps** [reply]
As far as I can tell, this is available for all service tiers when using a postgres datastore in heroku

https://devcenter.heroku.com/articles/heroku-data-links.

I will throw up some repro steps in a bit for configuring this locally.

**Reply by jkepps** [reply]
### Setup some dummy databases
```
createdb test-a
createdb test-b
psql test-a
```

```sql
CREATE TABLE users (
  id serial not null constraint users_pkey primary key,
  "name" varchar
);

ALTER TABLE users OWNER TO [YOUR_DB_USERNAME];

INSERT INTO users (name) VALUES ('John Doe');
\q
```

```
psql test-b
```

```sql
CREATE TABLE cars (
  id serial not null constraint cars_pkey primary key,
  "make" varchar,
  "model" varchar,
  "user_id" integer not null
);

ALTER TABLE cars OWNER TO <YOUR DB USERNAME>;

INSERT INTO cars (make, model, user_id) VALUES ('Tesla', 'Model S', 1);
\q
```

### Create the data link
```
psql test-a
```

```sql
CREATE EXTENSION IF NOT EXISTS postgres_fdw;

CREATE SERVER IF NOT EXISTS datalink_test_b
  FOREIGN DATA WRAPPER postgres_fdw
  OPTIONS (dbname 'test-b');

CREATE USER MAPPING FOR CURRENT_USER
  SERVER datalink_test_b
  OPTIONS (user '<YOUR DB USERNAME>');

CREATE SCHEMA datalink_test_b;

IMPORT FOREIGN SCHEMA public
  FROM SERVER datalink_test_b
  INTO datalink_test_b;
```

### Test it out
```sql
SELECT * FROM datalink_test_b.cars;

SELECT * FROM users
JOIN datalink_test_b.cars  cars ON (cars.user_id = users.id);
```

**Reply by jkepps** [reply]
I would expect an update statement such as this
```sql
UPDATE users
  SET name = model
  FROM datalink_test_b.cars cars
  WHERE cars.user_id = users.id;
```
to work in a prisma migration since it is valid SQL and updates the records appropriately when I run it in psql. But when i try to run something similar in a prisma migration, the error message I mentioned above is thrown.

**Reply by jkepps** [reply]
@janpio were you able to repro? This is currently blocking some data migrations. Any info you can give on what's going on under the hood would be great, since that'll help me determine if I need to think of a workaround.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
